src/Plugin/tmgmt/Translator/ThebigwordTranslator.php

```diff
@@ -0,0 +1,872 @@
+    $remote_mapping = reset($remote_mappings);
+    $project_id = $remote_mapping->getRemoteIdentifier2();
+
+    /** @var Url $url */
+    $uri = $job_item->getSourceUrl()->getInternalPath();
+    $state = 'ReferenceAdd';
+    $name = 'source-url';
+    if ($preview) {
+      $source_plugin = $job_item->getSourcePlugin();
+      if ($source_plugin instanceof SourcePreviewInterface) {
+        $url = $source_plugin->getPreviewUrl($job_item);
+        $uri = $url->getInternalPath() . '?key=' . $url->getOption('query')['key'];
+      }
+      else {
+        $uri = '';
+      }
+      $state = 'ResourcePreviewUrl';
+      $name = 'preview-url';
+    }
+    $preview_data = '<?xml version="1.0" encoding="UTF-8"?><!DOCTYPE PreviewUrl SYSTEM "http://www.thebigword.com/dtds/PreviewUrl.dtd">
+<PreviewUrl>' . $_SERVER['HTTP_HOST'] . '/' . $uri . '</PreviewUrl>';
+
+    /** @var \DateTime $required_by */
+    $required_by = $job_item->getJob()->getSetting('required_by');
+    $required_by->setTimezone(new \DateTimeZone('UTC'));
+    $datetime = $required_by->format('Y-m-d\TH:i:s');
+    $form_params = [
+      'ProjectId' => $project_id,
+      'RequiredByDateUtc' => $datetime,
+      'SourceLanguage' => $job_item->getJob()->getRemoteSourceLanguage(),
+      'TargetLanguage' => $job_item->getJob()->getRemoteTargetLanguage(),
+      'FilePathAndName' => "$name.xml",
+      'FileState' => $state,
+      'FileData' => base64_encode($preview_data),
+      'FileIdToUpdate' => $file_id,
+    ];
+    $file_id = $this->request('file', 'PUT', $form_params);
+
+    return $file_id;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function pullRemoteTranslations(Translator $translator) {
+    $translated = 0;
+    $untranslated = 0;
+    $result = $this->pullRemoteTranslationsForStatus($translator, 'TranslatableReviewPreview');
+    $translated += $result['translated'];
+    $untranslated += $result['untranslated'];
+    $result = $this->pullRemoteTranslationsForStatus($translator, 'TranslatableComplete');
+    $translated += $result['translated'];
+    $untranslated += $result['untranslated'];
+    return [
+      'updated' => $translated,
+      'non-updated' => $untranslated,
+    ];
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function pullRemoteTranslationsForStatus(Translator $translator, $status) {
+    $this->setTranslator($translator);
+    $result = RemoteMapping::loadByLocalData(NULL, NULL, 'tmgmt_thebigword');
+    $remotes = [];
+    /** @var RemoteMapping $remote */
+    foreach ($result as $remote) {
+      $remotes[$remote->getRemoteIdentifier2()] = $remote;
+    }
+    $translated = 0;
+    $not_translated = 0;
+
+    try {
+      $files = $this->requestFileinfo($status);
+      foreach ($files as $file) {
+        $file_id = $file['FileId'];
+        $project_id = $file['ProjectId'];
+        /** @var RemoteMapping $remote */
+        $remote = isset($remotes[$project_id]) ? $remotes[$project_id] : NULL;
```

Comment:
I think the optimization here is backwards. You're loading *all* mappings that were ever created, which could be thousands. To avoid loading per project id within the loop, but you very likely only have a handful of files here at that point, as we ,as mentioned by Mike, would "fake-confirm" anything we don't know anyway.

If you remove that, you also remove the need for the data item key mapping trick. And you could actually optimize this way better by loading all project ID's and then doing your own entity query for all project ID's of files that you need.